### ISSUE TITLE
(websocket)when the network is bad, it shows “Fatal read error on socket transport”.

### ISSUE THREAD

**Issue by yjqiang** [omz/Pythonista-Issues#563]
https://github.com/aio-libs/aiohttp/issues/3202

According to the authors of aiohttp, it is a bug in pythonista.

**Comment by cclauss** [comment]
Please change the title of this issue to something more descriptive.  There a lot of open issues and the current title does nothing to help those who are trying to find and fix issues.

**Comment by yjqiang** [comment]
https://github.com/yjqiang/bilibili-live-tools/blob/master/connect.py

https://github.com/yjqiang/bilibili-live-tools/blob/master/bilibiliCilent.py

**Comment by yjqiang** [comment]
I used "try" to catch the exceptions. But I can't catch the "fatal error" msg.

**Comment by dgelessus** [comment]
Isn't this the same issue that was discussed in #555?

**Comment by yjqiang** [comment]
Sorry about that, at first I thought it is aiohttp client's bug, but I wrote a websocket using aiohttp, and it showed again. And I tried https://github.com/aaugustin/websockets and aiohttp, it showed again and again.

**Comment by yjqiang** [comment]
I think it should be the bug in pythonista.

**Comment by jsbain** [comment]
So, can you summarize how we reproduce this error?  Seems like there should be a very succinct script with a few lines that reproduces the issue.  Does this fail when the network is off? Or does it need interaction with the server for this to fail

**Comment by yjqiang** [comment]
I use aiohttp to connect websites. When the network is not stable. It shows this. I used "try except" for all the script, but it did not work.

**Comment by jsbain** [comment]
Again, we need a minimal example, not hundreds of lines of code.

Have you tested your code with a pc and the same server?

**Comment by yjqiang** [comment]
[bug for pyista.zip](https://github.com/omz/Pythonista-Issues/files/2536658/bug.for.pyista.zip)

**Comment by jsbain** [comment]
Not a pythonista problem. I reproduced your error on Android, when turning off data while the script ran.

I suspect you could duplicate on PC if you unplugged Ethernet, etc after the connection started.

**Comment by yjqiang** [comment]
I wrapped almost every network code with 'try'. And aiohttp's author said it is not a bug on aiohttp. Now it is not a bug in pythonista. ...I'm almost crying.

**Comment by yjqiang** [comment]
Can you help me to debug? I have been suffering from it for a long time.

**Comment by jsbain** [comment]
https://github.com/python/asyncio/issues/135

A search for the error would have resulted in this documented behavior in asyncio.

**Comment by jsbain** [comment]
https://github.com/python/asyncio/issues/80
Solution seems to be a custom error hndler to silence the warning.

**Comment by jsbain** [comment]
Or handle connection_lost, and don't worry about the extra logged warning

**Comment by yjqiang** [comment]
I wrapped all function using aiohttp with "try". But it kept showing this. And it seems it is a feature in asyncio instead of a bug?

**Comment by yjqiang** [comment]
I just don't know the reason why it shows this. Because when the connections lost in low level like tcp, python should resend the msg itself. If the connection is broken for http connection, I used 'try' to catch it. So where is this bug coming from?

**Comment by yjqiang** [comment]
Or maybe aiohttp used asyncio to make a connection but aiohttp didn't handle this erro?

**Comment by cclauss** [comment]
Correct.  That is why there is the suggestion that install a custom error handler.

Your __try__ / __except__ is not working because it is not your code that is raising.  Instead, it is code in the aio event loop that is raising so the custom error handler should be the right path forward.

**Comment by yjqiang** [comment]
Maybe aiohttp should handle it instead of me? I don't think user should handle timeout error in this way. It is so strange.

**Comment by jsbain** [comment]
This error occurs when the server closes the connection unexpectedly or otherwise you are disconnected.  The error is logged, but calls connection_lost on the Transport or Protocol.  So if you want to handle the actual problem so you can retry, then implement a connection_lost handler.  But if you want to turn off logging, you need the custom error_handler.

This is coming from asyncio, not aiohttp or anywhere else.

**Comment by yjqiang** [comment]
But why it shows? If the network is not good, python can wait until I read the msg from it, and then it raises this error. It just shows in the background and I have to catch it out of the connection.

**Comment by jsbain** [comment]
try this:
``` 
from asyncio import constants
constants.LOG_THRESHOLD_FOR_CONNLOST_WRITES = 1000000  # some big number
``` 
The "error" you see is just a logging of an exception that ultimately will result in a different event that your code is handling.

**Comment by jsbain** [comment]
try this:
``` 
from asyncio import constants
constants.LOG_THRESHOLD_FOR_CONNLOST_WRITES = 1000000  # some big number
``` 
The "error" you see is just a logging of an exception that ultimately will result in a different event that your code is handling.

**Comment by jsbain** [comment]
Actually, it looks like if you don't want the exception handler called, you can add to _FATAL_ERROR_IGNORE:
    asyncio.base_events._FATAL_ERROR_IGNORE.append(TimeoutError)

Maybe go ahead and close the various issues you have open on this topic?

**Comment by yjqiang** [comment]
And I just keep one issue open here for this problem. Where are the others?

**Comment by cclauss** [comment]
https://github.com/omz/Pythonista-Issues/issues/created_by/yjqiang

**Comment by yjqiang** [comment]
I still didn't find the answers to these questions.  And these are not the same question.

**Comment by jsbain** [comment]
Your questions are not related to pythonista.  I'd suggest stackoverflow or general aiohttp/asyncio questions.

Re fatal error ignore, you are right about it being a tuple.  You would use 
   asyncio._FATAL_ERROR_IGNORE += (TimeoutError,)
to append to it.

**Comment by yjqiang** [comment]
Sorry. Thx a lot for this. Sorry to bother you for a long time. Thank you.
